src/design/econsecurity.md

```diff
@@ -1,13 +1,83 @@
+Any resources used by a transaction must explicitly be paid for, and it is a module author's job to ensure that appropriate fees are required. Maintaining the balance between **resources used** and **price paid** is an important design activity for runtime security.
 
 *Indeed, mispriced EVM operations have shown how operations that underestimate cost can open economic DOS attack vectors: [Onwards; Underpriced EVM Operations](https://www.parity.io/onwards/), [Under-Priced DOS Attacks on Ethereum](https://www4.comp.polyu.edu.hk/~csxluo/DoSEVM.pdf)*
 
-<!-- ## todo
 
-* existing transaction fee
 
-* more fee structure examples -->
\ No newline at end of file
+Substrate provides several ways to affect the fees charges for executing a transaction. Substrate developer hub contains full details about [fees](https://substrate.dev/docs/en/next/development/module/fees) and [weights](https://substrate.dev/docs/en/next/conceptual/runtime/weight).
+
+* Base fee - Applies a fixed fee to each and every transaction. A parameter in the `transaction_payment` module.
+
+* Length fee - Applies a fee proportional to the transaction's length in bytes. The constant is a parameter in the `transaction_payment` module.
+
+* Transaction weight - Each transaction can declare a weight, either fixed, or calculated from its parameters. This is exemplified briefly below and more thoroughly in the kitchen.
+
+* Weight to Fee - A function to convert weight to fee. It doesn't need to be linear, although it often is. The same conversion function is applied across all transactions from all modules in the runtime. This is exemplified briefly below and more thoroughly in the kitchen.
+
+
+## Assigning Transaction Weights
+
+For simple transactions a fixed weight will do.
+```rust
+decl_module! {
+	pub struct Module<T: Trait> for enum Call {
+
+		#[weight = SimpleDispatchInfo::FixedNormal(100)]
+		fn store_value(_origin, entry: u32) -> Result {
+			// --snip--
+		}
+```
+
+For more complex transactions, custom weight calculations can be performed.
+```rust
+pub struct Conditional(u32);
+
+impl WeighData<(&bool, &u32)> for Conditional {
+	fn weigh_data(&self, (switch, val): (&bool, &u32)) -> Weight {
+
+		if *switch {
+			val.saturating_mul(self.0)
+		}
+		else {
+			self.0
+		}
+	}
+}
+```
+
+These examples, and several others can be compiled in the kitchen.
+
+## Converting Weight To Fees
+
+In many cases converting weight to fees 1:1 will suffice and be accomplished with `ConvertInto`.
+```rust
+impl transaction_payment::Trait for Runtime {
+	// --snip--
+	type WeightToFee = QuadraticWeightToFee<WeightFeeConstant,
```

Comment:
missing a closing bracket and a few arguments?

Reply:
Oops, that one wasn't supposed to even be quadratic. Fixed now.